### PR TITLE
[S3 sensor] S3 Hook and sensor

### PR DESCRIPTION
**Pull request by artwr**
This PR is a first step to more advanced S3 operators.
The hook can currently check for buckets and keys, and will
parse boto, AWS sdks or s3cmd config files if they are defined
in the connection in the Airflow connection db.

The sensor uses the Hook to check for a particular file
(or key to follow amazon's nomenclature)

@mistercrunch

### CODE REVIEW

**Comment by artwr** [comment]
Hook tested and works on our infra. Might require a little more tweaking to be able to query s3 public datasets.

**Comment by mistercrunch** [comment]
Test plan? :)

File: airflow/operators/sensors.py [hunk 1]
```diff
@@ -226,3 +228,45 @@ def poke(self):
             return False
         print([i for i in f])
         return True
+
+
+class S3Sensor(BaseSensorOperator):
+    """
+    Waits for a file to be present in a S3 bucket
+    """
```
Comment (mistercrunch) [inline airflow/operators/sensors.py]: please put the field definition for the constructor here, I want all the operators to be well documented, by the way, can this wait for a folder to exist?

File: airflow/hooks/S3_hook.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+from airflow.hooks.base_hook import BaseHook
+
+import boto
+from boto.s3.connection import S3Connection
+boto.set_stream_logger('boto')
+logging.getLogger("boto").setLevel(logging.INFO)
+
+
+def _parse_s3_config(config_file_name, config_format='boto', profile=None):
```
Comment (mistercrunch) [inline airflow/hooks/S3_hook.py]: This could be @classmethod, but no it's fine as is.

**Comment by mistercrunch** [comment]
My main question is whether this can be used to sense for a folder, if not maybe it should? if so, let's clarify the operator's docstring

File: airflow/operators/sensors.py [hunk 3]
```diff
@@ -226,3 +228,45 @@ def poke(self):
...
+    template_fields = ('bucket_key', 'bucket_name')
+    __mapper_args__ = {
+        'polymorphic_identity': 'S3Sensor'
+    }
+
+    @apply_defaults
+    def __init__(
+            self, bucket_key,
+            bucket_name=None,
+            s3_conn_id=conf.get('hooks', 'S3_DEFAULT_CONN_ID'),
```
Comment (mistercrunch) [inline airflow/operators/sensors.py]: Maybe also add this key to the default config (airflow/configuration.py) and `initdb` in the command line `airflow/bin/cli.py`

File: airflow/operators/sensors.py (hunk 1 above)
Comment (artwr) [inline airflow/operators/sensors.py]: I will. I would have said yes, except the get_key is not consistent for folders : 
See for instance:
http://stackoverflow.com/questions/9954521/s3-boto-list-keys-sometimes-returns-directory-key

It depends on how the folder is created. I am unsure how to fix that off hand

**Comment by mistercrunch** [comment]
If it's complicated for folders just make it clear what the behavior is in the docs

**Comment by artwr** [comment]
I think I may have a solution with 
http://docs.aws.amazon.com/AmazonS3/latest/dev/ListingKeysHierarchy.html

I'll get back to you to make to operator more general

**Comment by mistercrunch** [comment]
Yes yes yes!
![img] (http://www.reactiongifs.com/wp-content/uploads/2013/03/thumbs-aloft.gif)

**Pull request merged**